platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+        this.apiClient = unavailableClient;
+        this.pingClient = unavailableClient;
+        this.realClient = null;
+        this.state = State.CLOSED;
+        this.future = null;
+        lock.notifyAll(); // wakeup those waiting for isAvailable(timeout)
+      }
+      finalizeStateChange(true, futureToCancel, previousClientToClose, false);
+    } finally {
+      listeners.clear();
+      initializers.clear();
+    }
+  }
+
+  @Override
+  public boolean isAvailable() {
+    final boolean available = (state == State.CONNECTED);
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(
+          "Solr({}): current availability is [{} = {}]",
+          core,
+          state,
+          SolrClientAdapter.availableToString(available));
+    }
+    if (available && wasNotRecent(lastPing, SolrClientAdapter.PING_MIN_FREQUENCY)) {
+      LOGGER.debug(
+          "Solr({}): Proxy is starting a background task to ping the client because the last ping was too long ago",
+          core);
+      executor.submit(this::backgroundPing);
+    }
+    return available;
+  }
+
+  @Override
+  public boolean isAvailable(long timeout, TimeUnit unit) throws InterruptedException {
+    Validate.notNull(unit, "invalid null time unit");
+    if (isAvailable()) { // quick check to avoid synchronization
+      return true;
+    }
+    // letting now be recomputed by the timedWait() method allows us to better control testing
+    long now = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
+    final long end = now + unit.toNanos(timeout);
+
+    synchronized (lock) {
+      while (true) {
+        if (state == State.CLOSED) {
+          return false;
+        }
+        final boolean available = isAvailable();
+
+        if (available) {
+          return true;
+        }
+        final long t = end - now;
+
+        if (t <= 0L) { // we timed out
+          return available;
+        }
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug(
+              "Solr({}): waiting {} to become available",
+              core,
+              DurationFormatUtils.formatDurationHMS(TimeUnit.NANOSECONDS.toMillis(t)));
+        }
+        now = waiter.timedWait(lock, now, t, TimeUnit.NANOSECONDS);
+      }
+    }
+  }
+
+  @Override
+  public boolean isAvailable(Listener listener) {
+    Validate.notNull(listener, "invalid null listener");
+    if (state != State.CLOSED) {
+      LOGGER.debug("Solr({}): registering a new listener [{}]", core, listener);
+      listeners.add(listener);
+    } // else - notify the listener at least once
+    final boolean available = isAvailable();
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(
+          "Solr({}): starting a background task to notify a new listener [{}] that the availability is [{}]",
+          core,
+          listener,
+          SolrClientAdapter.availableToString(available));
+    }
+    executor.submit(() -> notifyAvailability(listener, "is"));
```

Comment:
❓Threading out the calls to the listeners (and initializers) is safer, but makes the code and tests more complex. Did we have a strong case for doing this at this time? Could we make it synchronous now and if we need it in the future, extract some kind of `AsyncListenerList` class and let it handle the executor and async notifications?

Reply:
It ain't just safer. In my opinion, it is typically the number one reasons for thread locks and issues down the road. When a client can call your code, you should always be weary of calling it back.

Reply:
I understand the risks and tradeoffs here. I was just trying to see if we absolutely needed this right now or if we could add it a later time when we really needed it. This adds more complexity to the class and tests, so I was trying to see if we could just add it later.

Reply:
discussed and will leave it as is to protect from a listener being called from a user's thread calling into the API.